Add DAL-free compass calibration.

markshannon opened the pull request:
The PR breaks the dependence of the compass calibration of the DAL storage and display components.
Reduces the flash requirements by 3.6k.

This will allow us (in a future PR) to use a cut-down DAL freeing a further ~10k of flash and ~2k of RAM.

Removing the dependence on the DAL storage component removes the risk of the file system and DAL storage component overwriting each other's data.

The calibration method is different from the one in the DAL as I am unconvinced that the maths in that implementation is correct. The paper it refers explains how to map an offset spheroid onto a centred sphere, but the model assumes 100s if not 1000s of data points distributed all over the spheroid. 
I think that with 12 data points in a circle, it just doesn't work, especially if the normal axis of the circle happens to lie north-south.

My implementation only gathers 16 data points in a circle, but in a horizontal plane and centres the resulting ellipse on the mean point. This is much easier to reason about, and uses *much* less RAM.

Experimentally the results seem at least as good as the previous implementation, although accuracy is still poor relative to a proper compass.

Compass heading (at 90 degrees intervals, conforming µBit to desk edge).

|  Ground truth  |  Master  | This PR |
|--------|--------|--------|
|   6     | 326  (-40) |  329  (-37) |
|   96   |  44  (-48)  |   81 (-15)  |
|  186  |   98 (-88)  |  156 (-30) |
|  276  |  282 (+6)   |  240 (-30) |

dpgeorge commented on source/microbit/microbitcompass.cpp:
```diff
@@ -31,93 +31,227 @@ extern "C" {
+} compasss_calibration;
+
+typedef union _persistent_compasss_data {
+    int8_t padding[1024];
```
What is the reason for this padding?

dpgeorge commented on source/microbit/microbitcompass.cpp:
```diff
@@ -31,93 +31,227 @@ extern "C" {
-    uBit.display.disable();
-    uBit.systemTicker.detach();
-    ticker_start();
+    microbit_display_scroll(&microbit_display_obj, "Keep dot pointed ahead. Press button for each point.", false);
```
Could this msg be shortened so it doesn't take so long to read, eg "Keep dot ahead, press button for next"?

dpgeorge commented:
I tried the new calibration and it seems to work.  But it's not that obvious how to do the calibration: should the current dot face north?  Should you rotate in a plane or can the microbit be angled?  I think this change to the calibration method is a very big user-facing change that will throw many users off.

Is it possible to use the same interface as the DAL uses (drawing a circle using the display and accelerometer to guide you) but collect and average points as is done here?  That would be the best from a user's point of view.

Also, I see that is compass spelled with triple-s in some places :)

ntoll commented:
Morning! Apologies for the relative silence on my part - I've been ill. :-(

I like @dpgeorge's suggestion of keeping the "UI" the same (if possible) for your new calibration technique.

You wouldn't believe the amount of confusion this is likely to cause given the large number of non-technical users we have. Anything that's just slightly different is likely to cause consternation for such users and it may mean teachers having to re-write resources (a great way to make enemies of a teacher is to tell them they have to re-write their lesson plans and worksheets).

I realise from our PoV this is a relatively trivial change but for nervous and inexperienced teacher's perspectives, it's quite a radical and potentially pain/work-causing change.

markshannon commented:
I initially intended to replicate the original UI. That's how I discovered that it doesn't work :disappointed: 
I wish I'd found out *before* I'd implemented the Python matrix class.

I'm now looking into gathering sufficient samples off-line to enable a reference pre-calibration.
It won't work if the device is attached to a robot or other lump of metal, but that is a problematic scenario for a compass anyway.

@finneyj
@jaustin 
Given that each micro:bit is identical, and calibration is both inaccurate and a nuisance, what is the reason that it isn't pre-calibrated?

finneyj commented:
@markshannon @dpgeorge 

Each micro:bit sadly isn't identical in this regard. There's an unknown linear error in each axis of the magnetometer. This is not calibrated out when the sensor is manufactured as those folks assume that it'll be taken out by software calibration... so they don't bother. 

You're right of course that all micro:bits are the same, so any hard iron offsets caused by the PCB could be precomputed - but not the static offset in the sensor, which from what I've seen is the dominant source of error.

I used to have a simpler algorithm too - like the one you describe above - that works great on a plane, but not when tilted (as you need to start taking all three axes into account and using the accelerometer to detemine device posture... which brings further complexity). You're right that the current calibration algorithm I wrote is far from perfect. I lost my christmas to this last year to meet a BBC requirement to have tilt compensation... (trials showed that kids hold the micro:bit in all sorts of orientations and expect the compass to still track). 

This was the closest I could get without burning too much memory. Would be interesting to punt in more samples and see how much difference it makes against RAM footprint, then find a more optimal point (just because we have 12 LEDs doesn't mean we can't take more samples... )

markshannon commented:
Personally, I think accurate calibration of the µBit is next to impossible without some external reference.
If a pre-calibrated compass is no worse (or at least not significantly worse) than the current approach, then it would make the µBit a lot easier to use (and save us a fair bit of code duplication).

Pre-calibrating the µBit would also mean that `compass.heading()` would not have the side-effect of forcing human intervention.
Forcing human interaction for a simple method call like `compass.heading()` breaks just about every API design guideline. Imagine you've just bolted the µBit to the desk to prevent it being pocketed in a public place. Let's hope it is a light desk :muscle:
I believe I may have mentioned this before :wink: https://github.com/lancaster-university/microbit-dal/issues/82

@finneyj 
The current implementation works in a single plane as well. The difference is that the plane is now vertical, not horizontal.
Do you have any data regarding the relative accuracy of the two methods or what difference adding more samples makes?

carlosperate commented:
It might also be worth pointing out a comment from @blmorris from a previous conversation about calibration: https://github.com/bbcmicrobit/micropython/issues/166#issuecomment-175103060

jaustin commented:
Can we do both? Ship a built in 'best guess (possibly quite a bad best
guess)' calibration and have a call to calibrate the compass? That way we
even make a learning point out of it...

I'd like to know what @whaleygeek thinks of this: David do you imagine such
a scheme will generate constant support tickets along the lines of "My
compass is not accurate" -> "Did you calibrate it?" ->"No"-> "Whoops"?

What kind of feedback options do we have for this? in MicroPython are you
allowed to print 'diagnostic'/debug info to the REPL to be helpful, like
saying "Using uncalibrated compass, results will not be accurate if people
use the compass API without first calibrating"?

Given the existing DAL compass API, it looks like one could ship this
pre-calibrated data and call compass.setCalibration() before doing other
compassy things:
https://developer.mbed.org/teams/Lancaster-University/code/microbit-dal/docs/eb91bba49623/MicroBitCompass_8cpp_source.html#l00703

On the other hand, if this pre-defined calibration is junk, there's
probably no point in doing this...



On 20 December 2016 at 13:43, Carlos <notifications@github.com> wrote:

> It might also be worth pointing out a comment from @blmorris
> <https://github.com/blmorris> from a previous conversation about
> calibration: #166 (comment)
> <https://github.com/bbcmicrobit/micropython/issues/166#issuecomment-175103060>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/bbcmicrobit/micropython/pull/384#issuecomment-268245980>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAI-qVvXkoNjuwYeVvZ_NkMi62twNG4Dks5rJ9t4gaJpZM4LQLnL>
> .
>

deshipu commented:
The advantage of this approach is that then the calibrating script can
be an external program shipped with the examples, and thus be much more
complex and accurate than what is possible with a built-in function, and
also save some space.

On Tue, 20 Dec 2016 05:54:56 -0800
Jonathan Austin <notifications@github.com> wrote:

> Can we do both? Ship a built in 'best guess (possibly quite a bad best
> guess)' calibration and have a call to calibrate the compass? That
> way we even make a learning point out of it...
> 
> I'd like to know what @whaleygeek thinks of this: David do you
> imagine such a scheme will generate constant support tickets along
> the lines of "My compass is not accurate" -> "Did you calibrate it?"
> ->"No"-> "Whoops"?
> 
> What kind of feedback options do we have for this? in MicroPython are
> you allowed to print 'diagnostic'/debug info to the REPL to be
> helpful, like saying "Using uncalibrated compass, results will not be
> accurate if people use the compass API without first calibrating"?
> 
> Given the existing DAL compass API, it looks like one could ship this
> pre-calibrated data and call compass.setCalibration() before doing
> other compassy things:
> https://developer.mbed.org/teams/Lancaster-University/code/microbit-dal/docs/eb91bba49623/MicroBitCompass_8cpp_source.html#l00703
> 
> On the other hand, if this pre-defined calibration is junk, there's
> probably no point in doing this...
> 
> 
> 
> On 20 December 2016 at 13:43, Carlos <notifications@github.com> wrote:
> 
> > It might also be worth pointing out a comment from @blmorris
> > <https://github.com/blmorris> from a previous conversation about
> > calibration: #166 (comment)
> > <https://github.com/bbcmicrobit/micropython/issues/166#issuecomment-175103060>
> >
> > —
> > You are receiving this because you were mentioned.
> > Reply to this email directly, view it on GitHub
> > <https://github.com/bbcmicrobit/micropython/pull/384#issuecomment-268245980>,
> > or mute the thread
> > <https://github.com/notifications/unsubscribe-auth/AAI-qVvXkoNjuwYeVvZ_NkMi62twNG4Dks5rJ9t4gaJpZM4LQLnL>
> > .
> >  
> 
> 



-- 
Radomir Dopieralski

-- 
Radomir Dopieralski

jaustin commented:
Also... A passing thought. Maybe we should do some crowdsourced science on this - here's a quick strawman:
Generate a hex file that does the following
* Loads a pre-computed calibration
* Asks the user to perform some defined set of rotations/tilts relative to a local starting origin, storing each time the recorded heading value
* Ask the user to perform the full calibration routine
* Ask the user to perform the same set of rotations/tilts relative to the same local origin. Store the heading values as above.

Print on the screen and over UART the following, asking the user to record it and tell us:
* Some combined metric of correlation between the two sets of measurements (IE how close was the heading with default calibration and with the device-specific calibration?)
* The final 'calibration data' that is determined.

This gives us an idea of how bad the pre-configured offsets would be, and also perhaps helps us get a better set of pre-computed offsets by taking a wider range of environments into account than we could do with a few of us at our desks. It does make the assumption that the calibration step is effective in getting closer to a 'real result' but I don't think that's in question?

I'd suggest we ask people in the slack channel (http://tech.microbit.org/get-involved/where-to-find/) and MicroPython mailing list to do the experiment. 

The more advanced version of this might also use a separate device with GPS or a (shock!) real compass to provide some absolute values into the mix...

whaleygeek commented:
Well, what I can say is that if you give people a user guide, they won't read it. We have details on the website already about compass calibration, and we still get support tickets saying 'can't understand how to calibrate the compass'. So it's got to be blindingly obvious at the point of use, to be of any use at all.

The best possible outcome would be for the first-experience hex file that comes pre-loaded, to do a device specific compass calibration as part of the 'game', and for the persistence to last across re-flashes. That way it is a 'do this when you get it out of the box' experience, rather than a 'I just bolted it to a table leg and now I have to spin the table' experience.

finneyj commented:
@markshannon Oh, I think I might well recall our previous github discussion on the matter. :-) You're quite right of course, and let's just say that it wasn't by choice that the calibration step became a push feature... 

Whilst I love the simplicity of a pre-computed calibration, I have real doubts it could work in practice. Simply because:

- we'd need to determine that calibration for each micro:bit (e.g. in the factory where we make the micro:bits) and burn it into FLASH somehow. This would add a lot of complexity and cost, a you'd probably need an additional rig for each micro:bit, a two stage programming pipeline, controlled magnetic conditions, etc... 

- I'm not sure this would work globally. Even if we *could* do the above, different parts of the word have different magnetic field strengths... This is also part of what gets calibrated out when you run a calibration algorithm. We'd probably need to ship a micro:bit round the world again to test this one out... which might be fun though. e.g. check out the map here, and the underlying key... https://planet-earth-2017.com/2014/06/23/current-magnetic-field-map-confirms-3-magnets-configuration/

If we really wanted to have a rough calibration that worked out of the box (which would be awesome if we could), we'd probably be better off either:

a) talking to Freescale to see if they can supply pre-trimmed MAG3110 modules for us.
b) look for an alternative magnetometer that is pre-trimmed, and use that instead.

Both these solutions only work for future devices of course... so we'd still have about 1M devices we'd need to support through an alternative too.

markshannon commented:
I've changed the underlying mechanism to provide `start_calibration` and `stop_calibration` methods. Calibration occurs in the background between those calls.

When running, calibration records the 26 most extreme values along the vectors (-1/0/-1),(-1/0/-1),(-1/0/-1) excepting (0,0,0). When complete, values that are insufficiently spaced are discarded and then the mean is taken.

Should `heading` get called when the compass is not calibrated then the calibration occurs in a blocking fashion, as it does now.

markshannon commented:
This fixes #408. Whether the underlying issue is fixed or not isn't clear, but as we no longer call into the DAL, it seems likely.

dpgeorge commented:
@markshannon I've tested this new version out but it hangs for me in the first loop in "distance".  You loop over bit but seem to decrease it twice for each loop, so it counts 15, 13, 11, 9, 7, 5, 3, 1, -1, -3, ... and never stops.

dpgeorge commented:
Also, I needed to decrease the heap by 8 bytes to get it to compile.

markshannon commented:
@dpgeorge I've fixed the double-decrement in the loop in "distance". 
Could you check if you still need to add 8 bytes? It compiles as-is on my machine.

dpgeorge commented on docs/compass.rst:
```diff
@@ -18,9 +18,9 @@ Functions
+    Starts the calibration process. A brief message will be scrolled
     to the user after which they will need to rotate the device in order to
-    draw a circle on the LED display.
+    keep the flashing pixel pointing in the same direction.
```
This doc should be unchanged because you still draw a circle.

markshannon commented:
Should be ready to go now.
I've reverted the comment in the docs, reduced the heap use by 8 bytes, and added code to raise a `RuntimeError` if the display is off when calibrating.

dpgeorge commented:
Merged!  Thanks @markshannon for the hard work!

whaleygeek commented:
Does this still say 'draw a circle' when the calibration sequence
initiates, and does the process involve tilting the microbit to fill in
dots on a circle and show a happy face in success?

On 30 Mar 2017 01:20, "Damien George" <notifications@github.com> wrote:

> Merged! Thanks @markshannon <https://github.com/markshannon> for the hard
> work!
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/bbcmicrobit/micropython/pull/384#issuecomment-290265212>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AD5cAZM4-kRECgw96RzZmmE1niFzMvW2ks5rqvVrgaJpZM4LQLnL>
> .
>

dpgeorge commented:
> Does this still say 'draw a circle' when the calibration sequence initiates, and does the process involve tilting the microbit to fill in dots on a circle and show a happy face in success?

Yes, yes, yes.  It's almost indistinguishable from the DAL version.

It also has persistence: once a given microbit is calibrated you don't need to calibrate it again (even after a power off).  But you can initiate a calibration (compass.calibrate()) at any time.
